### PR TITLE
Changes needed to expose FSM timeouts to clients

### PR DESCRIPTION
**Pull request by evanmcc**
- change delete in riak_client to honor timeouts passed
  in Options
- change the various object interfaces to collect and
  pass on the timeout values

requires https://github.com/basho/riak_pb/issues/38

### CODE REVIEW

**Comment by evanmcc** [comment]
associated riak test: https://github.com/basho/riak_test/pull/243

File: src/riak_client.erl [hunk 1]
```diff
@@ -250,7 +250,13 @@ delete(Bucket,Key,Options,Timeout) when is_list(Options) ->
     ReqId = mk_reqid(),
     riak_kv_delete_sup:start_delete(Node, [ReqId, Bucket, Key, Options, Timeout,
                                            Me, ClientId]),
-    wait_for_reqid(ReqId, Timeout);
+    RTimeout = recv_timeout(Options),
+    lager:info("timeouts r ~p t ~p > ~p",
```
Comment (seancribbs) [inline src/riak_client.erl]: This should be at debug level and probably have a more useful message.

File: src/riak_kv_wm_object.erl [hunk 2]
```diff
@@ -848,19 +873,27 @@ ensure_doc(Ctx=#ctx{doc=undefined, key=undefined}) ->
...
 
 %% @spec delete_resource(reqdata(), context()) -> {true, reqdata(), context()}
 %% @doc Delete the document specified.
 delete_resource(RD, Ctx=#ctx{bucket=B, key=K, client=C, rw=RW, r=R, w=W,
-        pr=PR, pw=PW, dw=DW, timeout=Timeout}) ->
-    Options = lists:filter(fun({_, default}) -> false; (_) -> true end,
-        [{rw, RW}, {r, R}, {w, W}, {pr, PR}, {pw, PW}, {dw, DW}, 
-         {timeout, Timeout}]),
+        pr=PR, pw=PW, dw=DW}) ->
+    Options0 = lists:filter(fun({_, default}) -> false; (_) -> true end,
```
Comment (seancribbs) [inline src/riak_kv_wm_object.erl]: Suggested refactoring:

``` erlang
Opts0 = [{rw, RW}, {r, R}, {w, W}, {pr, PR}, {pw, PW}, {dw, DW}, {timeout, Timeout}],
Options = [ {Opt, Val} || {Opt, Val} <- Opts0, Val /= undefined, Val /= default ]
```

That pattern might also be useful in these other places where you conditionally include the timeout option, e.g.

``` erlang
Opts0 ++ [ {timeout, Timeout} || Timeout /= undefined ]
```

File: src/riak_client.erl [hunk 3]
```diff
@@ -295,7 +301,11 @@ delete_vclock(Bucket,Key,VClock,Options,Timeout) when is_list(Options) ->
     ReqId = mk_reqid(),
     riak_kv_delete_sup:start_delete(Node, [ReqId, Bucket, Key, Options, Timeout,
                                            Me, ClientId, VClock]),
-    wait_for_reqid(ReqId, Timeout);
+    RTimeout = recv_timeout(Options),
+    case RTimeout > Timeout of
```
Comment (seancribbs) [inline src/riak_client.erl]: Why not:

``` erlang
wait_for_reqid(ReqId, erlang:max(Timeout, RTimeout))
```

File: src/riak_client.erl (hunk 1 above)
Comment (evanmcc) [inline src/riak_client.erl]: meant to remove.

File: src/riak_client.erl (hunk 3 above)
Comment (evanmcc) [inline src/riak_client.erl]: erlang:stdlib() > evan:brain()

File: src/riak_kv_wm_object.erl (hunk 2 above)
Comment (evanmcc) [inline src/riak_kv_wm_object.erl]: it's done in three different ways in three different places.  Would be happy to refactor it to be uniform, but wasn't sure that this PR was the right place.

Comment (seancribbs) [inline src/riak_kv_wm_object.erl]: I'm ok with adding refactoring all three, or simply making a function to call for the same effect. Cleaning up debt is always good.

Sean Cribbs

File: src/riak_kv_wm_object.erl [hunk 4]
```diff
@@ -245,29 +230,66 @@ malformed_request(RD, Ctx) when Ctx#ctx.method =:= 'POST'
...
+        TimeoutStr -> 
+            try
+                Timeout = list_to_integer(TimeoutStr),
+                {false, RD, Ctx#ctx{timeout=Timeout}}
+            catch
+                _:_ ->
+                    {true,
+                     wrq:append_to_resp_body(io_lib:format("Bad timeout "
+                                                           "value ~p, "
+                                                           "using ~p~n",
```
Comment (seancribbs) [inline src/riak_kv_wm_object.erl]: If you provide a bad timeout value, it doesn't actually use the default but returns a 400 response.

**Comment by seancribbs** [comment]
I also wonder about the arbitrary 100ms being added to the timeout, which produces some strange results if you set the timeout very low. In a sense, you cannot set a timeout lower than 100ms.

**Comment by seancribbs** [comment]
:+1: to merge.

**Pull request merged**